### PR TITLE
feat: support output.footer in plugins.

### PR DESCRIPTION
**Pull request by komyg**

### CODE REVIEW

**Comment by 7086cmd** [comment]
Thank you for your contribution! It looks like footers are now supported (see `crates/rolldown/src/ecmascript/format/iife.rs`, etc.). Please consider adding some tests in `crates/rolldown/tests/fixtures/misc` together with `banner` instead. Thanks again!

**Comment by komyg** [comment]
> Thank you for your contribution! It looks like footers are now supported (see `crates/rolldown/src/ecmascript/format/iife.rs`, etc.). Please consider adding some tests in `crates/rolldown/tests/fixtures/misc` together with `banner` instead. Thanks again!

Hi, are you sure they are implemented?

I've ran the test I added on this PR on `main`, and it failed:

```ts
import { expect, vi } from 'vitest'
import path from 'node:path'
import { defineTest } from '@tests'

const entry = path.join(__dirname, './main.js')

export default defineTest({
  config: {
    input: entry,
    plugins: [
      {
        name: 'test-plugin',
        footer: () => '/* Lorem ipsum */',
      },
    ],
  },
  afterTest: (output) => {
    expect(output.output[0].code).toContain('/* Lorem ipsum */')
  },
})
```

```
packages/rolldown test: ⎯⎯⎯⎯⎯⎯⎯ Failed Tests 1 ⎯⎯⎯⎯⎯⎯⎯
packages/rolldown test:  FAIL  tests/fixture.test.ts > plugin/footer
packages/rolldown test: Error: Failed in ./fixtures/plugin/footer/_config.ts
packages/rolldown test:  ❯ tests/fixture.test.ts:27:15
packages/rolldown test:      25|         }
packages/rolldown test:      26|       } catch (err) {
packages/rolldown test:      27|         throw new Error(`Failed in ${testConfigPath}`, { cause: err })
packages/rolldown test:        |               ^
packages/rolldown test:      28|       }
packages/rolldown test:      29|     })
packages/rolldown test: Caused by: AssertionError: expected '\n//#region main.js\nconsole.log();\n…' to contain '/* Lorem ipsum */'
packages/rolldown test:  ❯ Object.afterTest tests/fixtures/plugin/footer/_config.ts:18:35
packages/rolldown test:  ❯ tests/fixture.test.ts:24:28
packages/rolldown test: ⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯
packages/rolldown test: Serialized Error: { operator: 'strictEqual' }
packages/rolldown test: ⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯⎯[1/1]⎯
packages/rolldown test:  Test Files  1 failed | 2 passed (3)
packages/rolldown test:       Tests  1 failed | 54 passed | 1 skipped (56)
packages/rolldown test:    Start at  07:26:25
packages/rolldown test:    Duration  1.03s (transform 203ms, setup 0ms, collect 1.21s, tests 876ms, environment 0ms, prepare 121ms)
packages/rolldown test: Failed
```

**Comment by 7086cmd** [comment]
I am sorry for making this mistake, ignoring the footer in the plugins.

File: crates/rolldown/src/ecmascript/ecma_generator.rs [hunk 1]
```diff
@@ -73,11 +73,22 @@ impl Generator for EcmaGenerator {
...
+        Some(footer) => footer.call(&rendered_chunk).await?,
+        None => None,
+      };
+      ctx
+        .plugin_driver
+        .footer(HookFooterArgs { chunk: &rendered_chunk }, footer.unwrap_or_default())
+        .await?
     };
 
+    // let footer = match ctx.options.footer.as_ref() {
```
Comment (7086cmd) [inline crates/rolldown/src/ecmascript/ecma_generator.rs]: Just remove them. cc @underfin, I think it's Ok.
Thank you a lot!

**Review by 7086cmd — APPROVED** [review]

File: crates/rolldown/src/ecmascript/ecma_generator.rs (hunk 1 above)
Comment (komyg) [reply crates/rolldown/src/ecmascript/ecma_generator.rs]: I just removed, it.

**Comment by komyg** [comment]
> I am sorry for making this mistake, ignoring the footer in the plugins.

No worries, thank you for reviewing :)

**Review by 7086cmd — APPROVED** [review]
Thanks a lot.

**Comment by 7086cmd** [comment]
Note that node tests are not passed yet, you may modify it.

Remember to change the title. Suggestion: feat: support output.footer in plugins.

**Comment by 7086cmd** [comment]
I have modified your code to ensure it passes the tests. However, I don't have write access to make the changes directly. The test failures were due to the expected trimming of the final newline, not incorrect code behavior. Can I open another PR? The commit after squashed will also include you.

**Comment by komyg** [comment]
> I have modified your code to ensure it passes the tests. However, I don't have write access to make the changes directly. The test failures were due to the expected trimming of the final newline, not incorrect code behavior. Can I open another PR? The commit after squashed will also include you.

Sure, I think you can either open a PR targeting this branch or a completely new one with my commits.

Anyway, I think maintainers can make changes to the PRs, I know @underfin did it on my last one...

**Comment by komyg** [comment]
PS: thank you for fixing this

**Comment by 7086cmd** [comment]
> > I have modified your code to ensure it passes the tests. However, I don't have write access to make the changes directly. The test failures were due to the expected trimming of the final newline, not incorrect code behavior. Can I open another PR? The commit after squashed will also include you.
> 
> Sure, I think you can either open a PR targeting this branch or a completely new one with my commits.
> 
> Anyway, I think maintainers can make changes to the PRs, I know @underfin did it on my last one...

It is too annoying to fix this (just removing extra new lines and modifying file name validation targets) :(. I faced the same issue when refactoring it, so maybe I am familiar with this annoying test result :). Anyway, I am still not a maintainer yet.

**Pull request closed without merging**